// redux/commit/reducer.js
import t from './types'

const ACTION_HANDLERS = {
  [t.SET_COMMITS_FILTER]: (state, { filter }) => ({
    ...state,
    filter,
  }),
  [t.SET_CURRENT_USER]: (state, { user }) => ({
    ...state,
    user,
  }),
  [t.SET_CURRENT_REPO]: (state, { repo }) => ({
    ...state,
    activeRepo: repo,
  }),
  [t.GET_REPOS_REQUEST]: state => ({
    ...state,
    isLoadingRepos: true,
  }),
  [t.GET_REPOS_SUCCESS]: (state, { repos: { data } }) => ({
    ...state,
    isLoadingRepos: false,
    repos: data,
  }),
  [t.GET_REPOS_ERROR]: (state, { error }) => ({
    ...state,
    isLoadingRepos: false,
    error,
  }),
  [t.GET_COMMITS_REQUEST]: (state, { repo, description }) => ({
    ...state,
    activeRepo: repo,
    activeRepoDescription: description,
    isLoadingCommits: true,
  }),
  [t.GET_COMMITS_SUCCESS]: (state, { commits: { data } }) => {
    const commits = data.map(({ commit: { message, committer: { name, date } }, html_url }) => {
      const formattedDate = new Date(date)
      return {
        message,
        author_name: name,
        commit_url: html_url,
        date: formattedDate.toString(),
      }
    })
    return {
      ...state,
      isLoadingCommits: false,
      commits,
    }
  },
  [t.GET_COMMITS_ERROR]: (state, { error }) => ({
    ...state,
    isLoadingCommits: false,
    error,
  }),
}

const initialState = {
  repos: [],
  commits: [],
  user: '',
  error: '',
  filter: '',
  activeRepo: null,
  isLoadingRepos: true,
  isLoadingCommits: false,
  activeRepoDescription: '',
}

export default function reducer(state = initialState, action) {
  const handler = ACTION_HANDLERS[action.type]
  return handler ? handler(state, action) : state
}
