// components/Common/Search/index.jsx
import React, { Fragment } from 'react'
import PropTypes from 'prop-types'
import { Field } from 'redux-form'
import './style.css'

const renderField = (arg) => {
  const { input, disabled, className, type, placeholder } = arg
  return (
    <input
      type={type}
      disabled={disabled}
      className={className}
      placeholder={placeholder}
      {...input}
    />
  )
}

const defaultHandleSubmit = e => {
  e.preventDefault()
}

const UserSearch = ({
  handleSubmit,
  pristine,
  submitting,
  name,
  placeholder,
  disabled,
  className,
}) =>
  <Fragment>
    <form className="userSearch__inputWrapper" onSubmit={handleSubmit || defaultHandleSubmit}>
      <Field
        name={name}
        type="text"
        disabled={disabled}
        component={renderField}
        placeholder={placeholder}
        className={className || 'userSearch__input'}
      />
      {handleSubmit && <div className="userSearch__submit">
        <button
          type="submit"
          disabled={pristine || submitting}>Search</button>
      </div>}
    </form>
  </Fragment>

UserSearch.propTypes = {
  name: PropTypes.string.isRequired,
  className: PropTypes.string,
  placeholder: PropTypes.string.isRequired,
  pristine: PropTypes.bool,
  submitting: PropTypes.bool,
  disabled: PropTypes.bool,
  handleSubmit: PropTypes.func,
}

export default UserSearch
