// components/User/Profile/Panel/index.jsx
import React, { PureComponent } from 'react'
import { List } from 'react-virtualized'
import PropTypes from 'prop-types'
// import lodash from 'lodash'
import { v1 } from 'uuid'
import { getDimension } from 'utils'
import './style.css'

/* eslint-disable jsx-a11y/heading-has-content */
const renderLoaderEl = () => (
  <div key={v1()} className="panel_item loading">
    <h5 className="loading__background profile__message" />
    <h5 className="loading__background profile__url" />
    <h5 className="loading__background profile__author" />
    <h5 className="loading__background profile__date" />
  </div>
)
/* eslint-disable jsx-a11y/heading-has-content */

class Panel extends PureComponent {
  state = {
    width: getDimension('width', 200, 200),
    rowHeight: getDimension('height', 200, 200),
    loaders: {
      repos: true,
      followers: true,
      following: true,
      stars: true,
    },
  }

  resizeDimensions = () => {
    this.setState({
      width: getDimension('width', 200, 200),
      rowHeight: getDimension('height', 200, 200),
    })
  }

  static getDerivedStateFromProps({ isLoadingRepos, isLoadingFollowers, isLoadingFollowing, isLoadingStars }) {
    return {
      loaders: {
        repos: isLoadingRepos,
        followers: isLoadingFollowers,
        following: isLoadingFollowing,
        stars: isLoadingStars,
      },
    }
  }

  componentDidMount() {
    window.addEventListener('resize', this.resizeDimensions)
  }

  componentWillUnmount() {
    window.removeEventListener('resize', this.resizeDimensions)
  }

  renderLoaders = () => {
    let i = 5
    const loaders = []
    while (i > 0) {
      loaders.push(renderLoaderEl())
      i -= 1
    }
    return loaders
  }

  renderRepos = () => {
    const { width, rowHeight } = this.state
    const { repos } = this.props
  
    const rowRenderer = ({ index, key, style }) => {
      const { name, html_url, created_at, description } = repos[index]
      return (
        <div key={key} style={style} className="panel_item">
          <h5 className="profile__message">{name}</h5>
          <h5 className="profile__url">
            <a href={html_url} target="_blank">Url</a>
          </h5>
          <h5 className="profile__author">{description}</h5>
          <h5 className="profile__date">{(new Date(created_at).toString())}</h5>
      </div>
      )
    }
    return (
      <List
        width={width}
        rowHeight={rowHeight}
        style={{ outline: 0 }}
        rowCount={repos.length}
        rowRenderer={rowRenderer}
        height={rowHeight * Math.min(20, repos.length)}
      />
    )
  }

  renderData = tab => {
    if (tab === 'repos') {
      return this.renderRepos()
    } else if (tab === 'followers') {
      return this.renderRepos()
    } else if (tab === 'following') {
      return this.renderRepos()
    } else if (tab === 'stars') {
      return this.renderRepos()
    }
  }

  render() {
    const { activeTab } = this.props
    const { loaders } = this.state
    return (
      <div className="userProfile__panel">
        <div className="userProfilePanelItems__wrapper">
          {loaders[activeTab] ? this.renderLoaders() : this.renderData(activeTab)}
        </div>
      </div>
    )
  }
}

Panel.propTypes = {
  activeTab: PropTypes.string,
  repos: PropTypes.array,
  followers: PropTypes.array,
  following: PropTypes.array,
  stars: PropTypes.array,
  isLoadingRepos: PropTypes.boolean,
  isLoadingFollowers: PropTypes.boolean,
  isLoadingFollowing: PropTypes.boolean,
  isLoadingStars: PropTypes.boolean,
}

export default Panel
