// components/User/Profile/Header/index.jsx
import React from 'react'
import PropTypes from 'prop-types'
import './style.css'

const TABS = [
  {
    tab: 'repos',
    title: 'Repositories',
  },
  {
    tab: 'stars',
    title: 'Stars',
  },
  {
    tab: 'followers',
    title: 'Followers',
  },
  {
    tab: 'following',
    title: 'Following',
  },
]

const HeaderWrapper = ({ goBack, setActiveProfileTab, counts, activeTab }) => {

  return (
    <div className="profile_header__wrapper">
      <div onClick={goBack} className="header__backbutton" />
      <div className="header__items__wrapper">
        {TABS.map(({ tab, title }) => (
          <p className={`${tab === activeTab ? 'active' : ''}`} key={tab} onClick={() => setActiveProfileTab(tab)}>
            {title} <span className="count">{counts[tab]}</span>
          </p>
        ))}
      </div>

    </div>
  )
}

HeaderWrapper.propTypes = {
  activeTab: PropTypes.string.isRequired,
  counts: PropTypes.object.isRequired,
  goBack: PropTypes.func.isRequired,
  setActiveProfileTab: PropTypes.func.isRequired,
}

export default HeaderWrapper
